src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionIterator.java

```diff
@@ -0,0 +1,226 @@
+
+        Utils.nonNull(readShard);
+        Utils.nonNull(readHeader);
+        Utils.nonNull(evaluator);
+        Utils.validateArg(minRegionSize >= 1, "minRegionSize must be >= 1");
+        Utils.validateArg(maxRegionSize >= 1, "maxRegionSize must be >= 1");
+        Utils.validateArg(minRegionSize <= maxRegionSize, "minRegionSize must be <= maxRegionSize");
+        Utils.validateArg(assemblyRegionPadding >= 0, "assemblyRegionPadding must be >= 0");
+        Utils.validateArg(activeProbThreshold >= 0.0, "activeProbThreshold must be >= 0.0");
+        Utils.validateArg(maxProbPropagationDistance >= 0, "maxProbPropagationDistance must be >= 0");
+
+        this.readShard = readShard;
+        this.readHeader = readHeader;
+        this.reference = reference;
+        this.features = features;
+        this.evaluator = evaluator;
+        this.minRegionSize = minRegionSize;
+        this.maxRegionSize = maxRegionSize;
+        this.assemblyRegionPadding = assemblyRegionPadding;
+        this.activeProbThreshold = activeProbThreshold;
+        this.maxProbPropagationDistance = maxProbPropagationDistance;
+
+        this.readyRegion = null;
+        this.previousRegion = null;
+        this.pendingRegions = new ArrayDeque<>();
+        this.readCachingIterator = new ReadCachingIterator(readShard.iterator());
+        this.readCache = new ArrayDeque<>();
+        this.activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readHeader);
+
+        // We wrap our LocusIteratorByState inside an AllLocusIterator so that we get empty loci
+        // for uncovered locations. This is critical for reproducing GATK 3.x behavior!
+        final LocusIteratorByState libs = new LocusIteratorByState(readCachingIterator, DownsamplingMethod.NONE, false, ReadUtils.getSamplesFromHeader(readHeader), readHeader, false);
+        this.locusIterator = new AllLocusIterator(readShard.getInterval(), libs);
+
+        readyRegion = loadNextAssemblyRegion();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return readyRegion != null;
+    }
+
+    @Override
+    public AssemblyRegion next() {
+        if ( ! hasNext() ) {
+            throw new NoSuchElementException("next() called when there were no more elements");
+        }
+
+        previousRegion = readyRegion;
+        readyRegion = loadNextAssemblyRegion();
+        return previousRegion;
+    }
+
+    private AssemblyRegion loadNextAssemblyRegion() {
+        AssemblyRegion nextRegion = null;
+
+        while ( locusIterator.hasNext() && nextRegion == null ) {
+            final AlignmentContext pileup = locusIterator.next();
+
+            // Ignore sites that are not contained within this shard's interval
+            if ( ! readShard.getInterval().contains(pileup) ) {
+                continue;
+            }
```

Comment:
Theoretically unnecessary because the AllLocusIterator already asserts this.

Reply:
True, but it doesn't hurt.